cpp/src/shuffler/chunk.cpp

```diff
@@ -13,126 +13,243 @@
 namespace rapidsmpf::shuffler::detail {
 
 Chunk::Chunk(
-    PartID pid,
-    ChunkID cid,
-    std::size_t expected_num_chunks,
-    std::size_t gpu_data_size,
+    ChunkID chunk_id,
+    size_t n_messages,
+    std::vector<PartID> part_ids,
+    std::vector<size_t> expected_num_chunks,
+    std::vector<uint32_t> meta_offsets,
+    std::vector<uint64_t> data_offsets,
     std::unique_ptr<std::vector<uint8_t>> metadata,
-    std::unique_ptr<Buffer> gpu_data
+    std::unique_ptr<Buffer> data
 )
-    : pid{pid},
-      cid{cid},
-      expected_num_chunks{expected_num_chunks},
-      gpu_data_size{gpu_data_size},
-      metadata{std::move(metadata)},
-      gpu_data{std::move(gpu_data)} {}
+    : chunk_id_{chunk_id},
+      n_messages_{n_messages},
+      part_ids_{std::move(part_ids)},
+      expected_num_chunks_{std::move(expected_num_chunks)},
+      meta_offsets_{std::move(meta_offsets)},
+      data_offsets_{std::move(data_offsets)},
+      metadata_{std::move(metadata)},
+      data_{std::move(data)} {}
```

Comment:
*suggestions*: Let us do all the validation of the format of the `Chunk` here. Then we don't need assertions scattered around the rest of the code, since it will not be possible to construct an invalid chunk.